bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/FileRecoveryLog.java

```diff
@@ -0,0 +1,237 @@
+        if (oldFile.exists()) {
+            existingLogs = readLogsFromFile(oldFile);
+            if (deleteOldLogs) {
+                File[] files = recoveryLogDir.toFile().listFiles(
+                        (dir, name) -> name.matches(baseFileName + "(\\d+)\\.log")
+                );
+                for (File file: files) {
+                        file.delete();
+                    }
+                }
+            }
+        File newFile = recoveryLogDir.resolve(baseFileName + (latestVersion + 1) + ".log").toFile();
+        try {
+            Files.createDirectories(recoveryLogDir); // create directory if not exists
+            newFile.createNewFile();
+            initAppendChannel(newFile);
+            if (existingLogs == null) {
+                return newFile;
+            }
+            // write existing unfinished logs to the new file
+            if (!existingLogs.isEmpty()) {
+                cleanUpFinishedLogs();
+                if (!existingLogs.isEmpty()) {
+                    for (Map.Entry<String, TransactionLogRecord> entry : existingLogs.entrySet()) {
+                        put(entry.getValue());
+                    }
+                    existingLogs.clear();
+                }
+            }
+        } catch (IOException e) {
+            stderr.println("error: failed to create recovery log file in " + recoveryLogDir);
+        }
+        return newFile;
+    }
+
+    /**
+     * Finds the latest version of the recovery log file.
+     *
+     * @return The latest version of the recovery log file.
+     */
+    private int findLatestVersion() {
+        int latestVersion = 0;
+        File directory = recoveryLogDir.toFile();
+        File[] files = directory.listFiles((dir, name) -> name.matches(baseFileName + "(\\d+)\\.log"));
+        if (files == null) {
+            return latestVersion;
+        }
+        for (File file : files) {
+            String fileName = file.getName();
+            int version = Integer.parseInt(fileName.replaceAll(
+                    baseFileName, "").replaceAll(".log", "")
+            );
+            if (version > latestVersion) {
+                latestVersion = version;
+            }
+        }
+        return latestVersion;
+    }
+
+    /**
+     * Initializes the append channel for the given file.
+     *
+     * @param file The file to initialize the append channel for.
+     */
+    private void initAppendChannel(File file) {
+        try {
+            appendChannel = FileChannel.open(file.toPath(), StandardOpenOption.APPEND);
+            FileLock lock = appendChannel.tryLock();
+            if (lock == null) {
+                stderr.println("error: failed to acquire lock on recovery log file " + file.toPath());
+            }
+        } catch (IOException e) {
+            stderr.println("error: failed to acquire lock on recovery log file " + file.toPath());
+        }
+    }
+
+    @Override
+    public void put(TransactionLogRecord trxRecord) {
```

Comment:
Let's document all public functions with documentation comments

Reply:
Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/42080/commits/301950353ba46fa7df492b8ade559088735ad0a9.